File: elixir/prime-factors/lib/prime_factors.ex
defmodule PrimeFactors do
  @doc """
  Compute the prime factors for 'number'.

  The prime factors are prime numbers that when multiplied give the desired
  number.

  The prime factors of 'number' will be ordered lowest to highest.
  """
  @spec factors_for(pos_integer) :: [pos_integer]
  def factors_for(number) do
    number
    |> :math.sqrt()
    |> floor()
    |> primes_to()
    |> Enum.filter(fn x -> rem(number, x) == 0 end)
    |> then(fn divers -> factorate(number, divers) end)
  end

  defp factorate(number, candidates, factors \\ [])

  defp factorate(number, _, factors) when number <= 1,
    do: factors |> Enum.reverse()

  defp factorate(number, [], factors) when number > 1,
    do: [number | factors] |> Enum.reverse()

  defp factorate(_, [], factors),
    do: factors |> Enum.reverse()

  defp factorate(number, [c | _] = cands, factors) when rem(number, c) == 0,
    do: factorate(div(number, c), cands, [c | factors])

  defp factorate(number, [_ | tail], factors),
    do: factorate(number, tail, factors)

  defp primes_to(number) when number < 2, do: []
  defp primes_to(2), do: [2]
  defp primes_to(3), do: [2, 3]

  defp primes_to(number) do
    previous_primes = number |> :math.sqrt() |> floor() |> primes_to()
    last = previous_primes |> List.last()

    previous_primes
    |> Enum.concat(
      max(last, 3)..number//2
      |> Enum.reject(fn c ->
        previous_primes
        |> Enum.any?(fn p -> rem(c, p) == 0 end)
      end)
    )
  end
end

File: cpp/prime-factors/prime_factors.cpp
#include "prime_factors.h"


namespace prime_factors {

std::vector<int> candidates(const int number) {
    std::vector<int> numbers;

    for(int i = 2; i <= number; i++) {
        numbers.push_back(i);
    }

    return numbers;
}

std::vector<int> primes_to(const std::vector<int> numbers, const int prime) {
    std::vector<int> primes;

    for(const int number: numbers) {
        if(number == prime || number % prime != 0) {
            primes.push_back(number);
        }
    }

    return primes;
}

std::vector<int> primes_at(const int value) {
    std::vector<int> primes = candidates(value);
    
    for(size_t i = 0; i < primes.size(); i++) {
        primes = primes_to(primes, primes[i]);
    }

    return primes;
}

bool is_prime(const int number) {
    if(number <= 1) {
        return false;
    }

    for(const int prime: primes_at(std::sqrt(number))) {
        if(number % prime == 0) {
            return false;
        }
    }

    return true;
}

std::vector<int> of(const int number) {
    const int no_factor = -1;
    
    if(number < 2) {
        return {};
    }

    if(is_prime(number)) {
        return {number};
    }

    int factor = no_factor;
    for(const int prime: primes_at(std::sqrt(number))) {
        if(number % prime == 0) {
            factor = prime;
            break;
        }
    }
    
    if(factor == no_factor) {
        return {};
    }

    std::vector<int> factors = of(number / factor);
    factors.push_back(factor);
    std::sort(factors.begin(), factors.end());

    return factors;
}


}  // namespace prime_factors
